hw/ip/otbn/dv/rig/rig/gens/known_wdr.py

```diff
@@ -0,0 +1,109 @@
+from ..program import ProgInsn, Program
+
+from ..snippet import ProgSnippet
+from ..snippet_gen import GenCont, GenRet, SnippetGen
+
+
+class KnownWDR(SnippetGen):
+    '''A snippet generator that generates known values (all zeros or all ones
+    for now) for WDRs.
+
+    '''
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        if 'bn.xor' not in insns_file.mnemonic_to_insn:
+            raise RuntimeError('BN.XOR instruction not in instructions file')
+
+        if 'bn.not' not in insns_file.mnemonic_to_insn:
+            raise RuntimeError('BN.NOT instruction not in instructions file')
+
+        self.insn_xor = insns_file.mnemonic_to_insn['bn.xor']
+        self.insn_not = insns_file.mnemonic_to_insn['bn.not']
+        # BN.XOR has six operands: wrd, wrs1, wrs2, shift_type, shift_value
+        # and flag_group
+        if not (len(self.insn_xor.operands) == 6 and
+                isinstance(self.insn_xor.operands[0].op_type, RegOperandType) and
+                self.insn_xor.operands[0].op_type.reg_type == 'wdr' and
+                self.insn_xor.operands[0].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[1].op_type, RegOperandType) and
+                self.insn_xor.operands[1].op_type.reg_type == 'wdr' and
+                not self.insn_xor.operands[1].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[2].op_type, RegOperandType) and
+                self.insn_xor.operands[2].op_type.reg_type == 'wdr' and
+                not self.insn_xor.operands[2].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[4].op_type, ImmOperandType)):
+            raise RuntimeError('BN.XOR instruction from instructions file is '
+                               'not the shape expected by the SmallVal generator.')
+
+        self.wrd_op_type = self.insn_xor.operands[0].op_type
+        self.wrs1_op_type = self.insn_xor.operands[1].op_type
+        self.imm_op_type = self.insn_xor.operands[4].op_type
+        self.imm_op_type.shift = 3
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 3:
+            return None
+
+        # The bn.xor-bn.not pair takes 2 instructions
+        if program.space < 2:
+            return None
+
+        if model.fuel < 2:
+            return None
+
+        # Pick wrd any old way: we can write to any register. This should
+        # always succeed.
+        wrd_val_xor = model.pick_reg_operand_value(self.wrd_op_type)
+        assert wrd_val_xor is not None
+
+        # Pick an operand value.
+        wrs1_val_xor = model.pick_reg_operand_value(self.wrs1_op_type)
+        assert wrs1_val_xor is not None
```

Comment:
This one can definitely fail (what if this is the first generator to run, when no WDR has a defined value?)